crates/primitives/src/col_list.rs

```diff
@@ -0,0 +1,554 @@
+#![forbid(unsafe_op_in_unsafe_fn)]
+
+extern crate alloc;
+
+use crate::ColId;
+use alloc::alloc::{alloc, dealloc, handle_alloc_error, realloc};
+use core::{
+    alloc::Layout,
+    cmp::Ordering,
+    fmt,
+    hash::{Hash, Hasher},
+    iter,
+    mem::{size_of, ManuallyDrop},
+    ops::Deref,
+    ptr::NonNull,
+    slice::from_raw_parts,
+};
+use either::Either;
+
+/// Constructs a `ColList` like so `col_list![0, 2]`.
+///
+/// A head element is required.
+/// Mostly provided for testing.
+#[macro_export]
+macro_rules! col_list {
+    ($head:expr $(, $elem:expr)* $(,)?) => {{
+        let mut list = $crate::ColList::new($head.into());
+        $(list.push($elem.into());)*
+        list
+    }};
+}
+
+/// An error signalling that a `ColList` was empty.
+#[derive(Debug)]
+pub struct EmptyColListError;
+
+/// A builder for a [`ColList`] making sure a non-empty one is built.
+pub struct ColListBuilder {
+    /// The in-progress list.
+    list: ColList,
+}
+
+impl Default for ColListBuilder {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl ColListBuilder {
+    /// Returns an empty builder.
+    pub fn new() -> Self {
+        Self {
+            list: ColList::from_inline(0),
+        }
+    }
+
+    /// Push a [`ColId`] to the list.
+    pub fn push(&mut self, col: ColId) {
+        self.list.push(col);
+    }
+
+    /// Build the [`ColList`] or error if it would have been empty.
+    pub fn build(self) -> Result<ColList, EmptyColListError> {
```

Comment:
Oh, and this is why a separate `ColListBuilder` is needed, forgot to comment on this. 👍